### ISSUE TITLE
Capability to disable some system services

### ISSUE THREAD

**Issue by rlex** [k0sproject/k0s#427]
**Is your feature request related to a problem? Please describe.**
Not sure if this is feature request or bug report. I wanted to create absolute bare minimum setup (control node, couple of workers, with nothing to deploy from start except CoreDNS).  
I tried creating cluster with following config
```
apiVersion: k0s.k0sproject.io/v1beta1
kind: Cluster
metadata:
  name: k0s
spec:
  api:
    address: host_ip_address
    sans:
    - host_ip_address
    - host_fqdn
    extraArgs: {}
  controllerManager:
    extraArgs: {}
  scheduler:
    extraArgs: {}
  storage:
    type: etcd
    kine: null
    etcd:
      peerAddress: host_ip_address
  network:
    podCIDR: 10.244.0.0/16
    serviceCIDR: 10.96.0.0/12
    provider: custom
  podSecurityPolicy:
    defaultPolicy: 00-k0s-privileged
  workerProfiles: []
extensions: null
images:
  konnectivity:
    image: us.gcr.io/k8s-artifacts-prod/kas-network-proxy/proxy-agent
    version: v0.0.13
  coredns:
    image: docker.io/coredns/coredns
    version: 1.7.0
  repository: ""
telemetry:
  interval: 10m0s
  enabled: true
```

I was expecting setup will only end up with konnectivity and coredns, however metrics-server and kube-proxy was still here

**Describe the solution you'd like**
Ability to disable stock pods

**Comment by jnummelin** [comment]
> Ability to disable stock pods

Well, as you said it's bit between a bug report and feature req. 😄  I'll mark it as feature req as it is currently working "as designed", although one can argue if that's a good design or not.


> kube-proxy  was still here

Why would you want to take kube-proxy out? That'll break some critical things on networking side.

> metrics-server was still here

This we could pretty easily make configurable through the yaml. something like:
```
metrics-server:
  disabled: true
```

I guess the images part of the documentation needs some TLC. The image section just allows one to point k0s to use images from non-standard registries eg. in case you want to pull the images from your own private registry.

**Comment by rlex** [comment]
>Why would you want to take kube-proxy out? That'll break some critical things on networking side.

Calico with ebpf / kube-router / etc. Of course it can be always disabled afterwards.

>This we could pretty easily make configurable through the yaml. something like

That will be great indeed.

**Comment by unixfox** [comment]
> This we could pretty easily make configurable through the yaml. something like:
> 
> ```
> metrics-server:
>   disabled: true
> ```
> 
> I guess the images part of the documentation needs some TLC. The image section just allows one to point k0s to use images from non-standard registries eg. in case you want to pull the images from your own private registry.

Could this work instead this way: if one omit a component from `images` then this component doesn't get loaded by k0s? I guess that's how rlex tried the first time and how I would think it would work too.

**Comment by jnummelin** [comment]
> if one omit a component from images then this component doesn't get loaded by k0s?

That would mean every user has to always define ALL the images, even if using the default ones. IMO that's too much friction for most users as ~95% cases would be using the default upstream image locations.

**Comment by unixfox** [comment]
> > if one omit a component from images then this component doesn't get loaded by k0s?
> 
> That would mean every user has to always define ALL the images, even if using the default ones

Isn't that already the case when using `k0s default-config`? I don't quite understand your point on that because most of the users will just use the default config generator and never touch the `images:` part from the config file.

**Comment by jnummelin** [comment]
> Isn't that already the case when using k0s default-config? I don't quite understand your point on that because most of the users will just use the default config generator and never touch the images: part from the config file.

Not really 100% same, one can run `k0s server` without ANY config file which means k0s uses the default config internally. (the exact same one sees with `k0s default-config`). In similar fashion, when some config "block" is missing entirely, k0s uses the defaults for that. So say we omit the `images` part totally from the config, k0s will use the default config for images.

This is done to reduce friction for users, in most cases users should probably configure only few options vs. having users exposed to all hundred options every time.

**Comment by mmack** [comment]
A option to disable kube-proxy would be awesome. To the best of my knowledge k0s would be the only k8s deployment tool next to plain k8s (kubespray for example) who allows deployments of cilium in kube-proxy free mode: https://docs.cilium.io/en/v1.9/gettingstarted/kubeproxy-free/

**Comment by unixfox** [comment]
I would really like to be able to deactivate the default coredns because I need to add custom hosts to it, but currently I'm not able to do that.

I tried to remove coredns manually but it came back after a restart of k0s.

**Comment by jnummelin** [comment]
Implemented in #1023
